[WGSL] Nested array types are not packed correctly

tadeuzagallo opened the pull request:
#### 24dad92a328381c12020352cdc88040a7ce90ec7
<pre>
[WGSL] Nested array types are not packed correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=288558">https://bugs.webkit.org/show_bug.cgi?id=288558</a>
<a href="https://rdar.apple.com/145584310">rdar://145584310</a>

Reviewed by Mike Wyrzykowski.

This originally landed in 291162@main, but hit an existing bug where we didn&apos;t
correctly unpack vec3&apos;s inside arrays. The fix is to unpack the base of a field
access when it is a vector.

The logic for packing array types in the GlobalVariableRewriter didn&apos;t
handle nested array types, so types like `array&lt;array&lt;T&gt;&gt;` wouldn&apos;t be
packed, even if `T` required packing. The code could also easily be
simplified by removing the custom logic under `packArrayType` and using
the generic `packType` function instead, which makes it correctly handle
the nested types.

* LayoutTests/fast/webgpu/nocrash/fuzz-145584310-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-145584310.html: Added.
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::packArrayResource):
(WGSL::RewriteGlobalVariables::packType):
(WGSL::RewriteGlobalVariables::packArrayType):

Canonical link: <a href="https://commits.webkit.org/291791@main">https://commits.webkit.org/291791@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/821dc90cd1e23e41bbf3977ce06b468c34da0570

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93980 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13565 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98988 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44507 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13863 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21997 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71713 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29063 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96982 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10294 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84882 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionAPIDevTools.MessagePassingFromPanelToBackground (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52053 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9975 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2543 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43823 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80229 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2625 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101029 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21032 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15323 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80722 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21284 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80867 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80084 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19958 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24631 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1989 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14192 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21016 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26194 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20703 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24163 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22444 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (commented):
I think this breaks  https://threejs.org/examples/?q=webgpu#webgpu_compute_particles

tadeuzagallo committed:
[WGSL] Nested array types are not packed correctly
https://bugs.webkit.org/show_bug.cgi?id=288558
rdar://145584310

Reviewed by Mike Wyrzykowski.

This originally landed in 291162@main, but hit an existing bug where we didn't
correctly unpack vec3's inside arrays. The fix is to unpack the base of a field
access when it is a vector.

The logic for packing array types in the GlobalVariableRewriter didn't
handle nested array types, so types like `array<array<T>>` wouldn't be
packed, even if `T` required packing. The code could also easily be
simplified by removing the custom logic under `packArrayType` and using
the generic `packType` function instead, which makes it correctly handle
the nested types.

* LayoutTests/fast/webgpu/nocrash/fuzz-145584310-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-145584310.html: Added.
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::packArrayResource):
(WGSL::RewriteGlobalVariables::packType):
(WGSL::RewriteGlobalVariables::packArrayType):

Canonical link: https://commits.webkit.org/291791@main
